### ISSUE TITLE
How to use input_mask for CustomQKVToContextPluginDynamic plugin?

### ISSUE THREAD

**Issue by 1049451037** [NVIDIA/TensorRT#2655]
## Description

I can run without input_mask right now. But when I add input_mask, trtexec will raise `Error[9]: [pluginV2Builder.cpp::reportPluginError::23] Error Code 9: Internal Error (/CustomQKVToContextPluginDynamic: could not find any supported formats consistent with input/output data types)`.

## Environment

nvidia docker container 22.12

## Relevant Files

https://github.com/NVIDIA/TensorRT/tree/release/8.5/plugin/bertQKVToContextPlugin

## Steps To Reproduce

Here is the code to generate onnx:

```python
import torch
import torch.nn as nn

# https://github.com/NVIDIA/TensorRT/tree/release/8.5/plugin/bertQKVToContextPlugin
# The yaml file says that version 3 is not supported yet.

class CustomQKVToContextPluginDynamic(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, i_mask, hidden_size, num_heads):

        return input
    @staticmethod
    def symbolic(g, input, i_mask, hidden_size, num_heads):
        return g.op("CustomQKVToContextPluginDynamic", input, i_mask, plugin_version_s='1', type_id_i=0, hidden_size_i=hidden_size, num_heads_i=num_heads, has_mask_i=True)

class MyModule(nn.Module):
    def __init__(self, hidden_size, num_heads):
        super().__init__()
        assert hidden_size % num_heads == 0
        self.hidden_size = hidden_size
        self.num_heads = num_heads
        self.size_per_head = hidden_size // num_heads
        self.Wq = nn.Linear(self.hidden_size, self.hidden_size)
        self.Wk = nn.Linear(self.hidden_size, self.hidden_size)
        self.Wv = nn.Linear(self.hidden_size, self.hidden_size)
    def forward(self, x, i_mask):
        # shape of x (seq_len, batch_size, hidden_size)
        # shape of i_mask (batch_size)
        # output (seq_len, batch_size, hidden_size)
        Q = self.Wq(x)
        K = self.Wk(x)
        V = self.Wv(x)
        qkv = torch.cat([Q, K, V], dim=2)
        qkv = qkv.view(x.size(0), x.size(1), 3, self.num_heads, self.size_per_head)
        qkv = qkv.transpose(2, 3).contiguous().view(x.size(0), x.size(1), 3*self.hidden_size, 1, 1)
        return CustomQKVToContextPluginDynamic.apply(qkv, i_mask, self.hidden_size, self.num_heads).select(-1, 0).select(-1, 0)

model = MyModule(768, 8).cuda()#.half()
input = torch.randn(128, 2, 768).cuda()#.half()
i_mask = torch.tensor([[64], [78]], dtype=torch.int32).cuda()

from torch.onnx import OperatorExportTypes
torch.onnx.export(model, (input, i_mask), 'test.onnx', operator_export_type=OperatorExportTypes.ONNX_FALLTHROUGH, input_names=['input_0', 'input_1'], output_names=['output_0'])

```

Then use `trtexec --onnx=test.onnx --saveEngine=test.trt` to generate engine.

**Comment by 1049451037** [comment]
@ttyio

**Comment by ttyio** [comment]
the mask is FP32 according to the code https://github.com/NVIDIA/TensorRT/blob/main/plugin/bertQKVToContextPlugin/qkvToContextPlugin.cpp#L243

@1049451037 could you modify your code? thx!

**Comment by 1049451037** [comment]
I have tried fp32 and it can do inference normally. But in the readme it says mask is an interger with int32 type. So I'm worried about the inference result...

**Comment by ttyio** [comment]
@1049451037 , I checked https://github.com/NVIDIA/TensorRT/tree/main/plugin/bertQKVToContextPlugin, there is no mention that `mask` should be int32. However the shape of the `mask` is incorrect, the shape should be [B, 1] for unfused version and [B, XMMAS_M * THREADS_PER_CTA] for fused version.

**Comment by 1049451037** [comment]
As the bertQKVToContextPlugin README says, the format of input_mask is same as that in embLayerNormPlugin. The embLayerNormPlugin [README](https://github.com/NVIDIA/TensorRT/tree/main/plugin/embLayerNormPlugin) says, it's int32.

By the way, what is XMMAS_M * THREADS_PER_CTA.

**Comment by ttyio** [comment]
@1049451037 hmmm, good catch, will fix the doc. thx!

`XMMAS_M` and `THREADS_PER_CTA` are gpu architecutre depend, If you grep the plugin code, you can find the values, and they are used in the mha kernel implementation (which is not open sourced) and the embLayernorm plugin (which will reorder the mask in the correct layout)

**Comment by ttyio** [comment]
closing since no activitity for more than 3 weeks, thanks!

**Comment by lix19937** [comment]
> the mask is FP32 according to the code https://github.com/NVIDIA/TensorRT/blob/main/plugin/bertQKVToContextPlugin/qkvToContextPlugin.cpp#L243
> 
> @1049451037 could you modify your code? thx!

Now  https://github.com/NVIDIA/TensorRT/blob/release/8.6/plugin/bertQKVToContextPlugin/qkvToContextPlugin.cpp#L243  or  https://github.com/NVIDIA/TensorRT/blob/main/plugin/bertQKVToContextPlugin/qkvToContextPlugin.cpp#L243    the mask is int32

**Comment by ttyio** [comment]
@lix19937 , yes it is an optimization to save engine build time, we change to int32 to avoid trt from trying fp32 and fp16 reformat on this plugin in latest code.
